app/src/main/java/org/schabi/newpipe/player/gesture/PopupPlayerGestureListener.kt

```diff
@@ -0,0 +1,287 @@
+        if (DEBUG)
+            Log.d(TAG, "onSingleTapConfirmed() called with: e = [$e]")
+
+        if (isDoubleTapping)
+            return true
+        if (player.exoPlayerIsNull())
+            return false
+
+        onSingleTap()
+        return true
+    }
+
+    override fun onScroll(
+        initialEvent: MotionEvent,
+        movingEvent: MotionEvent,
+        distanceX: Float,
+        distanceY: Float
+    ): Boolean {
+
+        if (isResizing) {
+            return super.onScroll(initialEvent, movingEvent, distanceX, distanceY)
+        }
+
+        if (!isMoving) {
+            playerUi.closeOverlayBinding.closeButton.animate(true, 200)
+        }
+
+        isMoving = true
+
+        val diffX: Float = (movingEvent.rawX - initialEvent.rawX)
+        var posX: Float = (initialPopupX + diffX)
+        val diffY: Float = (movingEvent.rawY - initialEvent.rawY)
+        var posY: Float = (initialPopupY + diffY)
+
+        if (posX > playerUi.screenWidth - playerUi.popupLayoutParams.width) {
+            posX = (playerUi.screenWidth - playerUi.popupLayoutParams.width).toFloat()
+        } else if (posX < 0) {
+            posX = 0f
+        }
+
+        if (posY > playerUi.screenHeight - playerUi.popupLayoutParams.height) {
+            posY = (playerUi.screenHeight - playerUi.popupLayoutParams.height).toFloat()
+        } else if (posY < 0) {
+            posY = 0f
+        }
+
+        playerUi.popupLayoutParams.x = posX.toInt()
+        playerUi.popupLayoutParams.y = posY.toInt()
+
+        // -- Determine if the ClosingOverlayView (red X) has to be shown or hidden --
+        val showClosingOverlayView: Boolean = playerUi.isInsideClosingRadius(movingEvent)
+        // Check if an view is in expected state and if not animate it into the correct state
+        val expectedVisibility = if (showClosingOverlayView) View.VISIBLE else View.GONE
+        if (binding.closingOverlay.visibility != expectedVisibility) {
+            binding.closingOverlay.animate(showClosingOverlayView, 200)
+        }
+
+        playerUi.windowManager.updateViewLayout(binding.root, playerUi.popupLayoutParams)
+        return true
+    }
+
+    override fun getDisplayPortion(e: MotionEvent): DisplayPortion {
+        return when {
+            e.x < playerUi.popupLayoutParams.width / 3.0 -> DisplayPortion.LEFT
+            e.x > playerUi.popupLayoutParams.width * 2.0 / 3.0 -> DisplayPortion.RIGHT
+            else -> DisplayPortion.MIDDLE
+        }
+    }
+
+    override fun getDisplayHalfPortion(e: MotionEvent): DisplayPortion {
+        return when {
+            e.x < playerUi.popupLayoutParams.width / 2.0 -> DisplayPortion.LEFT_HALF
+            else -> DisplayPortion.RIGHT_HALF
+        }
+    }
+
+    companion object {
+        private val TAG = PopupPlayerGestureListener::class.java.simpleName
+        private val DEBUG = MainActivity.DEBUG
+        private const val TOSS_FLING_VELOCITY = 2500
```

Comment:
What is this for? Maybe a short jdoc comment?

Reply:
This is the commit that introduced it: 38b2ffd4504cefaa79885d2f6b90a9c129645445. From the PR (#810) description:

> Added popup player tossing when flung with weaker strength (useful for getting the popup out of keyboard's way).

I personally don't understand [the code](https://github.com/TeamNewPipe/NewPipe/blob/26099bb47b821978a35ee961d7ffaf2c4a89a393/app/src/main/java/org/schabi/newpipe/player/gesture/PopupPlayerGestureListener.kt#L173-L190) (it sets a velocity to a position... it does not make sense from a metrics point of view).
Btw, "toss" and "fling" mean roughly the same thing, I don't know why there is such a redundancy.